pkg/solana/txm/pendingtx.go

```diff
@@ -556,6 +567,129 @@ func (c *pendingTxContext) TrimFinalizedErroredTxs() int {
 	return len(expiredIDs)
 }
 
+func (c *pendingTxContext) GetSignatureInfo(sig solana.Signature) (txInfo, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	info, exists := c.sigToTxInfo[sig]
+	if !exists {
+		return txInfo{}, ErrSigDoesNotExist
+	}
+	return info, nil
+}
+
+func (c *pendingTxContext) OnReorg(sig solana.Signature) (pendingTx, error) {
+	// Acquire a read lock to check if the signature exists and needs to be reset
+	err := c.withReadLock(func() error {
+		// Check if the signature is still being tracked
+		info, exists := c.sigToTxInfo[sig]
+		if !exists {
+			return ErrSigDoesNotExist
+		}
+
+		// Check if the transaction is still in a non finalized/errored state
+		var broadcastedExists, confirmedExists bool
+		_, broadcastedExists = c.broadcastedProcessedTxs[info.id]
+		_, confirmedExists = c.confirmedTxs[info.id]
+		if !broadcastedExists && !confirmedExists {
+			return ErrTransactionNotFound
+		}
+		return nil
+	})
+	if err != nil {
+		// If transaction or sig are not found, return
+		return pendingTx{}, err
+	}
+
+	var pTx pendingTx
+	// Acquire a write lock to perform the state reset
+	_, err = c.withWriteLock(func() (string, error) {
+		// Retrieve sig and tx again inside the write lock
+		info, exists := c.sigToTxInfo[sig]
+		if !exists {
+			return "", ErrSigDoesNotExist
+		}
+
+		tx, broadcastedProcessedExists := c.broadcastedProcessedTxs[info.id]
+		if broadcastedProcessedExists {
+			pTx = tx
+		}
+		tx, confirmedExists := c.confirmedTxs[info.id]
+		if confirmedExists {
+			pTx = tx
+		}
+
+		if !broadcastedProcessedExists && !confirmedExists {
+			// transaction does not exist in any non finalized/errored maps
+			return "", ErrTransactionNotFound
+		}
+
+		// reset state to broadcasted and update the transaction in the broadcasted map
+		info.state, pTx.state = Broadcasted, Broadcasted
```

Comment:
I think this should still work but wanted to run an edge case by you. If the TXM is configured to poll more slowly for statuses, there's a chance a confirmed tx gets re-org'd but also processed before we poll for statuses again. We could still detect a re-org if a tx goes from confirmed to processed. But I think this should still work since we handle broadcasted and processed the same way. We could just wait for the next status poll to update it to processed or if it jumps to confirmed by the next time that's ok too

Reply:
do you think passing the `currentStatus` to this method and update directly to that state should be better?

Reply:
Tbh I think it might be better to just keep it this way for simplicity but maybe we can add a comment in case anyone else has the same question. I think it's ok if we always mark it as Broadcasted here. We can pick up that it's really `Processed` on the next cycle. I don't think this introduces any risk with the expiration logic either. We wouldn't falsely pick up this transaction for expiration since we make sure to check any changes in confirmation status before expiration detection.

Reply:
added comment to explain this